File: frontend/src/types/BorrowingRecord.ts
import type { User } from './User';
import type { Book } from './Book';

export const BorrowingStatus = {
    BORROWED: 'BORROWED',
    RETURNED: 'RETURNED',
    OVERDUE: 'OVERDUE'
} as const;

export type BorrowingStatus = (typeof BorrowingStatus)[keyof typeof BorrowingStatus];

export interface BorrowingRecord {
    id: number;
    user: User;
    userId?: number; // legacy or for request payload
    book: Book;
    borrowDate: string;
    dueDate: string;
    returnDate?: string;
    status: BorrowingStatus;
    lateFee?: number;
    guestBorrowerName?: string;
    returnCondition?: string;
    notes?: string;
}

File: backend/src/main/java/com/bookmanagement/entity/BorrowingRecord.java
package com.bookmanagement.entity;

import jakarta.persistence.*;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;

@Entity
@Table(name = "borrowing_records")
@Data
@NoArgsConstructor
@EntityListeners(AuditingEntityListener.class)
public class BorrowingRecord {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "book_id", nullable = false)
    @com.fasterxml.jackson.annotation.JsonIgnoreProperties({ "hibernateLazyInitializer", "handler" })
    private Book book;

    // Nullable if guest
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = true)
    @com.fasterxml.jackson.annotation.JsonIgnoreProperties({ "hibernateLazyInitializer", "handler" })
    private User user;

    private String guestBorrowerName; // For non-registered users
    private String guestEmail;
    private String guestPhone;

    private String returnCondition; // Good, Damaged, Lost

    @Column(length = 1000)
    private String notes;

    @Column(nullable = false)
    private LocalDateTime borrowDate;

    @Column(nullable = false)
    private LocalDateTime dueDate;

    private LocalDateTime returnDate;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private BorrowingStatus status;

    private Double lateFee = 0.0;

    @CreatedDate
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;
}
